File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.js
/**
 * @param {number[][]} events
 * @return {number}
 */
var maxTwoEvents = function(events) {
 events.sort((a, b) => a[1] - b[1]);

    const n = events.length;
    const endT = new Array(n);
    const best = new Array(n);

    for (let i = 0; i < n; i++) {
        endT[i] = events[i][1];
        best[i] = events[i][2];
        if (i > 0) best[i] = Math.max(best[i], best[i - 1]);
    }

    let ans = 0;

    for (let i = 0; i < n; i++) {
        const st = events[i][0];
        const val = events[i][2];

        let l = 0, r = n - 1, idx = -1;
        while (l <= r) {
            const mid = Math.floor((l + r) / 2);
            if (endT[mid] < st) {
                idx = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }

        if (idx !== -1) ans = Math.max(ans, val + best[idx]);
        ans = Math.max(ans, val);
    }

    return ans;
};

File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.cs
public class Solution {
    public int MaxTwoEvents(int[][] events) {
         int n = events.Length;

        Array.Sort(events, (a, b) => a[0].CompareTo(b[0]));

        int[] start = new int[n];
        int[] end = new int[n];
        int[] val = new int[n];

        for (int i = 0; i < n; i++)
        {
            start[i] = events[i][0];
            end[i] = events[i][1];
            val[i] = events[i][2];
        }

        int[] suffixMax = new int[n];
        suffixMax[n - 1] = val[n - 1];

        for (int i = n - 2; i >= 0; i--)
        {
            suffixMax[i] = Math.Max(val[i], suffixMax[i + 1]);
        }

        int ans = 0;

        for (int i = 0; i < n; i++)
        {
            int l = i + 1, r = n;

            while (l < r)
            {
                int m = l + ((r - l) >> 1);
                if (start[m] > end[i])
                    r = m;
                else
                    l = m + 1;
            }

            if (l < n)
            {
                ans = Math.Max(ans, val[i] + suffixMax[l]);
            }

            ans = Math.Max(ans, val[i]);
        }

        return ans;
    }
}